src/cli/main.cc

```diff
@@ -170,8 +197,27 @@ int main(int argc, char *argv[]) {
   }
 #endif
 
+  const bool use_snapshot = config.snapshot_dir != "";
+  if (use_snapshot) {
+    if (const auto s = CreateSnapshot(config, config.snapshot_dir); !s.IsOK()) {
+      LOG(ERROR) << "Failed to create snapshot: " << s.Msg();
+      return 1;
+    }
+    LOG(INFO) << "Starting server in read-only mode with snapshot dir: " << config.snapshot_dir;
+    config.db_dir = config.snapshot_dir + "/db";
+  }
+  const auto snapshot_exit = MakeScopeExit([use_snapshot, &config]() {
+    if (use_snapshot && !config.keep_snapshot) {
+      LOG(INFO) << "Removing snapshot dir: " << config.snapshot_dir;
+      std::filesystem::remove_all(config.snapshot_dir);
+    }
+  });
+
   engine::Storage storage(&config);
-  s = storage.Open();
+  const bool read_only = use_snapshot;
+  const DBOpenMode open_mode = read_only ? kDBOpenModeForReadOnly : kDBOpenModeDefault;
+  s = storage.Open(open_mode);
```

Comment:
Since now they are just normal options, I think we'd better to move these logic to config.cc or storage.cc, to make it more maintainable.

Reply:
Also the `CreateSnapshot` function, it need to be moved out of `cli/main.cc` since its funtionality is out of CLI scope.

Reply:
Thanks, this makes sense: this is done now